cpp/ql/src/Critical/UseAfterFree.ql

```diff
@@ -54,8 +45,22 @@ class UseAfterFreeReachability extends StackVariableReachability {
   override predicate isSink(ControlFlowNode node, StackVariable v) { isDerefExpr(node, v) }
 
   override predicate isBarrier(ControlFlowNode node, StackVariable v) {
-    definitionBarrier(v, node) or
+    definitionBarrier(v, node)
+    or
     isFreeExpr(node, v)
+    or
+    exists(GuardCondition guard, ControlFlowNode defNode, boolean testIsTrue, SsaDefinition def |
+      // `node` may use the definition provided by `defNode`
+      def.getAUse(v) = node and
+      def.getAnUltimateDefiningValue(v) = defNode and
+      // `guard` controls the execution of `node`
+      guard.controls(node.getBasicBlock(), testIsTrue) and
+      // An equivalent guard controls the execution of `defNode`
+      globalValueNumber(guard)
+          .getAnExpr()
+          .(GuardCondition)
+          .controls(defNode.getBasicBlock(), testIsTrue)
+    )
```

Comment:
It took me a while to figure out why this is (probably) the right thing to do.  The key is that `defNode` is the conditional `malloc` in your example in the PR description, though it may take other forms than a `malloc`, e.g.
```
free(data); // source
...
if(b) {
  data = something_else; // defNode
}
...
if(b) { // guard
  use(data); // node (GOOD)
}
```
What we seem to be doing is simulating a more sophisticated analysis that understands `data` being conditionally a freed pointer (if `!b`).  The assignment to `data` itself is a `definitionBarrier` but this needs to be extended in some way to apply to the second conditional block.